include/dlaf/matrix/matrix_ref.h

```diff
@@ -0,0 +1,175 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <dlaf/matrix/distribution.h>
+#include <dlaf/matrix/matrix.h>
+#include <dlaf/matrix/matrix_base.h>
+#include <dlaf/matrix/tile.h>
+#include <dlaf/types.h>
+
+namespace dlaf::matrix {
+/// A @c MatrixRef represents a sub-matrix of a @c Matrix.
+///
+/// The class has reference semantics, meaning accesses to a @c MatrixRef and
+/// it's corresponding @c Matrix are interleaved if calls to read/readwrite are
+/// interleaved. Access to a @c MatrixRef and its corresponding @c Matrix is not
+/// thread-safe. A @c MatrixRef must outlive its corresponding @c Matrix.
+template <class T, Device D>
+class MatrixRef;
+
+template <class T, Device D>
+class MatrixRef<const T, D> : public internal::MatrixBase {
+public:
+  static constexpr Device device = D;
+
+  using ElementType = T;
+  using TileType = Tile<ElementType, D>;
+  using ConstTileType = Tile<const ElementType, D>;
+  using TileDataType = internal::TileData<ElementType, D>;
+  using ReadOnlySenderType = ReadOnlyTileSender<T, D>;
+
+  /// Create a sub-matrix of @p mat with an @p offset and @p size.
+  ///
+  /// @param[in] mat is the input matrix,
+  /// @param[in] offset is the offset of the new matrix relative to the input matrix,
+  /// @param[in] size is the size of the new matrix relative to the offset,
+  /// @pre origin.isValid()
+  /// @pre size.isValid()
+  /// @pre origin + size <= mat.size()
+  MatrixRef(Matrix<const T, D>& mat, const GlobalElementIndex& offset, const GlobalElementSize& size)
+      : internal::MatrixBase(Distribution(mat.distribution(), offset, size)), mat_const_(mat),
+        offset_(offset) {}
+
+  // TODO: default, copy, move construction?
+  // - default: no, don't want empty MatrixRef
+  // - copy: implementable, still refer to the original matrix
+  // - move: implement as copy, i.e. still refer to original matrix?
+  MatrixRef() = delete;
```

Comment:
What should we allow here?

My preference would be to make `MatrixRef` behave like `std::reference_wrapper`, i.e.:
- no default constructor and thus no empty state
- regular constructor taking a `Matrix` etc.
- no move constructor since no empty state
- copy constructor

Preferences?

Reply:
I made it completely immovable to be conservative in the end because
1. It's not safe to pass copies into new tasks anyway and having it immovable _may_ avoid mistakes like that, and
2. One can always take a `MatrixRef<>&` where needed or create a new `MatrixRef`.
We can revisit this later if it turns out that it's more convenient to make it behave like `reference_wrapper` in some use cases.